### PR TITLE
ICU-20627 Adding alias locales to new structure in res_index.

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20627
- [x] Updated PR title and link in previous line to include Issue number
- [ ] Issue accepted
- [ ] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/data/BUILDRULES.py [hunk 1]
```diff
@@ -517,18 +516,18 @@ def generate_tree(
...
-        deprecates_xml = ET.parse(deprecates_xml_path)
-        for child in deprecates_xml.getroot():
-            if child.tag == "alias":
-                synthetic_locales.add(child.attrib["from"])
-            elif child.tag == "emptyLocale":
-                synthetic_locales.add(child.attrib["locale"])
-            else:
-                raise ValueError("Unknown tag in deprecates XML: %s" % child.tag)
+    # Exclude the deprecated locale variants; see ICU-20628
+    synthetic_locales = set([
```
Comment (srl295) [inline icu4c/source/data/BUILDRULES.py]: should be data driven, better to add something to the xml

File: icu4c/source/data/BUILDRULES.py [hunk 2]
```diff
@@ -517,18 +516,18 @@ def generate_tree(
     ]
 
     # Generate res_index file
-    synthetic_locales = set()
-    if not config.ignore_xml_deprecates:
```
Comment (srl295) [inline icu4c/source/data/BUILDRULES.py]: I hadn't seen this `ignore_xml_deprecates` before. Is v8 using it now?

Comment (sffc) [reply icu4c/source/data/BUILDRULES.py]: I added it in #642 and #646 (post-64) but I hadn't formally proposed it at the time, and now I'm bringing it up and we're figuring out what that feature should look like.

**Comment by sffc** [comment]
OK.  I changed the PR to add the aliases to their own section of the res_index file.  Another PR will come later with an API to expose it.

**Comment by FrankYFTang** [comment]
Could your write a simple description of, once your PR land, how will the caller change (flag or filter json) to control such behavior. Thanks.

**Comment by sffc** [comment]
> Could your write a simple description of, once your PR land, how will the caller change (flag or filter json) to control such behavior. Thanks.

This PR does not expose anything that the user can control; that would be in a follow-up PR that includes an API proposal.  I will not close the ticket until that step is done.

**Comment by srl295** [comment]
I support this direction and will review the exact code soon

sent from android

________________________________

File: icu4c/source/python/icutools/databuilder/request_types.py [hunk 3]
```diff
@@ -296,12 +297,18 @@ def __init__(self, **kwargs):
 
     def apply_file_filter(self, filter):
         i = 0
-        while i < len(self.input_files):
-            if filter.match(self.input_files[i]):
+        while i < len(self.installed_files):
+            if filter.match(self.installed_files[i]):
                 i += 1
                 continue
```
Comment (markusicu) [inline icu4c/source/python/icutools/databuilder/request_types.py]: Wouldn't "else" be clearer than "continue"?

File: icu4c/source/python/icutools/databuilder/request_types.py [hunk 4]
```diff
@@ -296,12 +297,18 @@ def __init__(self, **kwargs):
...
-            del self.input_files[i]
-        return i > 0
+            del self.installed_files[i]
+        j = 0
+        while j < len(self.alias_files):
+            if filter.match(self.alias_files[i]):
+                j += 1
+                continue
```
Comment (markusicu) [inline icu4c/source/python/icutools/databuilder/request_types.py]: ditto

File: icu4c/source/python/icutools/databuilder/request_types.py (hunk 3 above)
Comment (sffc) [reply icu4c/source/python/icutools/databuilder/request_types.py]: Done

File: icu4c/source/python/icutools/databuilder/request_types.py (hunk 4 above)
Comment (sffc) [reply icu4c/source/python/icutools/databuilder/request_types.py]: Done

**Comment by sffc** [comment]
Please also note: `locale_dependencies.data["aliases"]` originates from the icu-locale-deprecates.xml file, *not* the CLDR supplementalMetadata.  This is a decision made in the CLDR NewLdml2IcuConverter space.  I filed a bug to follow up on whether we want to make this list pull from CLDR supplementalMetadata instead:

https://unicode-org.atlassian.net/browse/CLDR-13078

File: icu4c/source/data/BUILDRULES.py [hunk 5]
```diff
@@ -517,26 +516,26 @@ def generate_tree(
...
-    index_input_files = []
+    # Exclude the deprecated locale variants and root; see ICU-20628
+    excluded_locales = set([
+        "ja_JP_TRADITIONAL",
+        "th_TH_TRADITIONAL",
+        "de_",
+        "de__PHONEBOOK",
+        "es_",
+        "es__TRADITIONAL",
+        "root",
```
Comment (srl295) [inline icu4c/source/data/BUILDRULES.py]: 1. could add metadata to the deprecates xml to drive this 
2. or, at least add a comment linking to [ICU-20628] suggesting the above

File: icu4c/source/data/BUILDRULES.py (hunk 1 above)
Comment (sffc) [reply icu4c/source/data/BUILDRULES.py]: Part of the mechanical purpose of this change is to *avoid* having to read any XML file in Python.  It's extremely unlikely that this list will grow, and hopefully these files will simply cease to exist at some point in the future.  There is a ticket linked in the comment above with more information if anyone ever works on this again.

File: icu4c/source/data/BUILDRULES.py (hunk 5 above)
Comment (sffc) [reply icu4c/source/data/BUILDRULES.py]: I expanded the comment a bit.  PTAL.

**Comment by sffc** [comment]
Ping @srl295

File: icu4c/source/data/BUILDRULES.py [hunk 6]
```diff
@@ -517,26 +516,28 @@ def generate_tree(
...
-        for child in deprecates_xml.getroot():
-            if child.tag == "alias":
-                synthetic_locales.add(child.attrib["from"])
-            elif child.tag == "emptyLocale":
-                synthetic_locales.add(child.attrib["locale"])
-            else:
-                raise ValueError("Unknown tag in deprecates XML: %s" % child.tag)
-    index_input_files = []
+    # Exclude the deprecated locale variants and root; see ICU-20628. This
+    # could be data-driven, but we do not want to perform I/O in this script
```
Comment (srl295) [inline icu4c/source/data/BUILDRULES.py]: ^ Why? Just curious.

Comment (sffc) [reply icu4c/source/data/BUILDRULES.py]: Discussed.

**Comment by sffc** [comment]
Squashed.

**Pull request merged**